checks/image/image_check.go

```diff
@@ -0,0 +1,148 @@
+func (pc *containerImageCheck) Groups() []string {
+	return []string{"image"}
+}
+
+// Description returns a detailed human-readable description of what this check
+// does.
+func (pc *containerImageCheck) Description() string {
+	return "Checks if there are pods with containers in privileged mode"
+}
+
+// Run runs this check on a set of Kubernetes objects. It can return warnings
+// (low-priority problems) and errors (high-priority problems) as well as an
+// error value indicating that the check failed to run.
+func (pc *containerImageCheck) Run(ctx context.Context, objects *kube.Objects) ([]checks.Diagnostic, error) {
+	var diagnostics []checks.Diagnostic
+	repos := ctx.Value(checks.Repositories)
+	validRepos := make(map[string]struct{})
+	if repos != nil {
+		repoArr, ok := repos.([]string)
+		if ok {
+			for _, r := range repoArr {
+				validRepos[r] = struct{}{}
+			}
+		}
+	}
+	for _, pod := range objects.Pods.Items {
+		diagnostics = append(diagnostics, pc.checkPod(validRepos, pod)...)
+	}
+
+	return diagnostics, nil
+}
+
+// checkPod checks if all pod images are well formed and
+// do not have any errors, or obvious mistakes.
+func (pc *containerImageCheck) checkPod(validRepos map[string]struct{}, pod corev1.Pod) []checks.Diagnostic {
+	var diagnostics []checks.Diagnostic
+	containers := append(pod.Spec.Containers, pod.Spec.InitContainers...)
+	for _, container := range containers {
+		named, err := checkImageError(container.Name, container.Image)
+		if err != nil {
+			diagnostics = append(diagnostics, checks.Diagnostic{
+				Check:    pc.Name(),
+				Severity: checks.Error,
+				Message:  err.Error(),
+				Kind:     checks.Pod,
+				Object:   &pod.ObjectMeta,
+				Owners:   pod.ObjectMeta.GetOwnerReferences(),
+			})
+		} else {
+			for _, e := range lintImage(named, validRepos, container.Name, container.Image) {
+				diagnostics = append(diagnostics, checks.Diagnostic{
+					Check:    pc.Name(),
+					Severity: checks.Warning,
+					Message:  e.Error(),
+					Kind:     checks.Pod,
+					Object:   &pod.ObjectMeta,
+					Owners:   pod.ObjectMeta.GetOwnerReferences(),
+				})
+			}
+		}
+	}
+	return diagnostics
+}
+
+func checkImageError(containerName, image string) (reference.Named, error) {
+	named, err := reference.ParseNormalizedNamed(image)
+	if err != nil {
+		return nil, fmt.Errorf("%q's container image, %q, is malformed: %w", containerName, image, err)
+	}
+	return named, nil
+}
+
+const dockerIO = "docker.io"
+
+var (
+	errEmptyDomain   = errors.New("empty domain")
+	errInvalidDomain = errors.New("invalid domain")
+	errNoSha         = errors.New("no sha")
+)
+
+func lintImage(named reference.Named, validRepos map[string]struct{}, containerName, image string) (warnings []error) {
+	domain := reference.Domain(named)
+	// the reference lib inserts "docker.io" into the image if it is not present,
+	// we remove it if we find that it isn't specified.
+	if domain == dockerIO && strings.Index(image, dockerIO) != 0 {
+		domain = ""
```

Comment:
append to warnings here and don't check if domain is blank again?

Reply:
This is to set the domain to empty if "docker.io" is added by `ParseNormalizedNamed`.  Thereafter an empty check is made.